### PR TITLE
#77: Add in pause after specific interval, "skip" silences 

### PR DESCRIPTION
**Pull request by jw1540**
PR adds a crude implementation on skipping silences. It basically checks when the average power for the currently playing buffer drops below an arbitrary value, and then increases the rate. Once the power returns above the value, the rate is reduced back to 1.0. 

Pausing after a time just runs a timer for that period and then calls pause. Perhaps it should also observe the playing status and remove itself if the audio file ends prior to the timer completing?

### CODE REVIEW

**Comment by tanhakabir** [comment]
Awesome thanks! I'll look into this tonight

**Comment by tanhakabir** [comment]
So something I want to make sure is that the player itself is pretty light and extensible in terms of manipulations being attached to the engine. Meaning I want the player at the core to be almost like AVAudioEngine that can stream.

But I also love your feature suggestions and think others would find super valuable! I'm looking now into extracting out these to be in their own namespace of "Extra Features" which wouldn't be as tightly coupled to the core engine source code but still give the same functionality. I'll co-commit these changes so that your changes exist in the adjacent namespace!

File: Source/Engine/AudioEngine.swift [hunk 1]
```diff
@@ -201,4 +203,52 @@ class AudioEngine: AudioEngineProtocol {
...
+                                               to: Int(buffer.frameLength),
+                                               by: buffer.stride).map { channelDataValue[$0] }
+
+            let rms = sqrt(channelDataValueArray.map { $0 * $0 }.reduce(0, +) / Float(buffer.frameLength))
+
+            let avgPower = 20 * log10(rms)
+
+            let meterLevel = self.scaledPower(power: avgPower)
+            if meterLevel < 0.6 {
+                self.changeRate(to: 1.5)
```
Comment (tanhakabir) [inline Source/Engine/AudioEngine.swift:238]: I don't want to use hardcoded 1.5 for speeding up to skip silences as they might have the audio rate set to 4 or something by default. 

There's 2 options here:
1. We change this to add to the current rate by some appropriate factor. I'm think this rate increase for skipping silences should be less at higher rates so we'd need to determine some sort of ratio.
2. We assume that at rates above some number, like 3, the skip silences feature wouldn't be noticeable/helpful and then we could just restrict the rate change to a lower range of original rates. This makes it easy to just add some constant to the current rate.

Comment (jw1540) [reply Source/Engine/AudioEngine.swift:238]: That's a really good point that I hadn't considered. I like both suggestions and think a mix of the two would be best? Applying some appropriate factor definitely seems the right way to go and if the rate increases beyond to a point where the skip would be negligible it just switches the tap off?

**Comment by jw1540** [comment]
@tanhakabir I wonder if it's worth pulling these out into two separate PR's? That way we can add a small feature increment whilst thinking a little more about how to approach skipping silences.

**Comment by tanhakabir** [comment]
@jw1540 yeah if you're able to separate them out that would be great!

**Comment by tanhakabir** [comment]
Hey @jw1540 what resources did you use for skipping silences? I want to document that in the code for future reference

**Comment by tanhakabir** [comment]
Asked for documentation in #81

**Pull request closed without merging**